packages/frontend/core/src/components/image-preview/index.tsx

```diff
@@ -16,54 +17,131 @@ import {
-import {
-  captionStyle,
-  cursorStyle,
-  dividerStyle,
-  imageBottomContainerStyle,
-  imagePreviewActionBarStyle,
-  imagePreviewBackgroundStyle,
-  imagePreviewModalCaptionStyle,
-  imagePreviewModalCenterStyle,
-  imagePreviewModalCloseButtonStyle,
-  imagePreviewModalContainerStyle,
-  imagePreviewModalStyle,
-  loaded,
-  scaleIndicatorButtonStyle,
-  unloaded,
-} from './index.css';
-import {
-  hasAnimationPlayedAtom,
-  previewBlockIdAtom,
-  previewblocksAtom,
-} from './index.jotai';
+import * as styles from './index.css';
 
 const filterImageBlock = (block: BlockModel): block is ImageBlockModel => {
   return block.flavour === 'affine:image';
 };
 
+function resolveMimeType(buffer: Uint8Array): string {
+  if (
+    buffer[0] === 0x47 &&
+    buffer[1] === 0x49 &&
+    buffer[2] === 0x46 &&
+    buffer[3] === 0x38
+  ) {
+    return 'image/gif';
+  } else if (
+    buffer[0] === 0x89 &&
+    buffer[1] === 0x50 &&
+    buffer[2] === 0x4e &&
+    buffer[3] === 0x47
+  ) {
+    return 'image/png';
+  } else if (
+    buffer[0] === 0xff &&
+    buffer[1] === 0xd8 &&
+    buffer[2] === 0xff &&
+    buffer[3] === 0xe0
+  ) {
+    return 'image/jpeg';
+  } else {
+    // unknown, fallback to png
+    console.error('unknown image type');
+    return 'image/png';
+  }
+}
+
+function copyImageToClipboard(image: HTMLImageElement) {
+  const canvas = document.createElement('canvas');
+  canvas.width = image.naturalWidth;
+  canvas.height = image.naturalHeight;
+  const context = canvas.getContext('2d');
+  if (!context) {
+    console.warn('Could not get canvas context');
+    return;
+  }
+  context.drawImage(image, 0, 0);
+  canvas.toBlob(blob => {
+    if (!blob) {
+      console.warn('Could not get blob');
+      return;
+    }
+    const dataUrl = URL.createObjectURL(blob);
+    navigator.clipboard
+      .write([new ClipboardItem({ 'image/png': blob })])
+      .then(() => {
+        console.log('Image copied to clipboard');
+        URL.revokeObjectURL(dataUrl);
+        toast('Copied to clipboard.');
+      })
+      .catch(error => {
+        console.error('Error copying image to clipboard', error);
+        URL.revokeObjectURL(dataUrl);
+      })
+      .finally(() => {
+        canvas.remove();
+      });
+  }, 'image/png');
+}
+
+function saveBufferToFile(buffer: Uint8Array, filename: string) {
+  const blob = new Blob([buffer], { type: resolveMimeType(buffer) });
+  const url = URL.createObjectURL(blob);
+  const a = document.createElement('a');
+  a.href = url;
+  a.download = filename;
+  document.body.append(a);
+  a.click();
+  a.remove();
+}
+
 export type ImagePreviewModalProps = {
   docCollection: DocCollection;
   pageId: string;
+  host: HTMLElement;
+};
+
+const ButtonWithTooltip = ({
+  icon,
+  tooltip,
+  disabled,
+  ...props
+}: PropsWithChildren<{
+  disabled?: boolean;
+  icon?: ReactElement;
+  tooltip: string;
+  onClick: () => void;
+  className?: string;
+}>) => {
+  const element = icon ? (
```

Comment:
this is stupid, why icon only mode is not supported by Button component by default

Reply:
Right now `IconButton` & `Button` are two different implementations. `<Button />` has icon prop as well but it serves different purpose.
I wrap it here for ease of use only for image preview toolbar.

Reply:
Indeed, let's create an issue and deal with it later. [AF-960](https://linear.app/affine-design/issue/AF-960/deprecate-iconbutton-and-use-button-instead)